### PR TITLE
Introducing Atomic Compare And Swap Support to X86

### PR DESCRIPTION
**Pull request by 0dvictor**
X86 CG now supports Atomic Compare And Swap.

Signed-off-by: Victor Ding <dvictor@ca.ibm.com>

### CODE REVIEW

**Comment by mstoodle** [comment]
Hi @0dvictor another one with only a little information here. What's the plan here?

**Comment by 0xdaryl** [comment]
In an effort to help clear the logjam of inactive PRs, I will close this WIP PR on Friday June 22 unless I hear otherwise.  Please continue to work on it in your private fork until it is ready for review and merging.

**Comment by 0dvictor** [comment]
@0xdaryl I apologize for the long delay of removing the WIP. This PR is actually ready for merging.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
...
+   cg->decReferenceCount(node->getChild(2));
+   return oldvalue;
+   }
+
+static TR::Register* inline64BitAtomicCompareAndMemoryUpdateOn32Bit(TR::Node* node, bool returnVaue, TR::CodeGenerator* cg)
+   {
+   TR_ASSERT(TR::Compiler->target.is32Bit(), "32-bit only");
+   auto address  = cg->evaluate(node->getChild(0));
+   auto oldvalue = cg->longClobberEvaluate(node->getChild(1));
+   auto newvalue = cg->evaluate(node->getChild(2));
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp]: Same nit as above with the use of `auto`.

**Comment by 0xdaryl** [comment]
My point about testing is that I would like to have some regular means of testing this in OMR CI builds (e.g., test cases).

**Comment by mstoodle** [comment]
@0dvictor ping. I'd appreciate it if you could spend some time helping us to get through these older PRs. Please address @0xdaryl 's review comments.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
...
 
-   node->setRegister(result);
+   node->setRegister(value);
    cg->decReferenceCount(node->getChild(0));
    cg->decReferenceCount(node->getChild(1));
-   return result;
+   return value;
+   }
+
+static TR::Register* inlineAtomicCompareAndMemoryUpdate(TR::Node* node, TR_X86OpCodes op, bool returnVaue, TR::CodeGenerator* cg)
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Fixed.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 3]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
...
 
-TR::Register* OMR::X86::TreeEvaluator::performSimpleAtomicMemoryUpdate(TR::Node* node, TR_X86OpCodes op, TR::CodeGenerator* cg)
+static TR::Register* inlineAtomicMemoryUpdate(TR::Node* node, TR_X86OpCodes op, TR::CodeGenerator* cg)
    {
    TR_ASSERT((!TR_X86OpCode(op).hasLongSource() && !TR_X86OpCode(op).hasLongTarget()) || TR::Compiler->target.is64Bit(), "64-bit instruction not supported on IA32");
-   TR::Register* address = cg->evaluate(node->getChild(0));
-   TR::Register* value   = cg->evaluate(node->getChild(1));
-   TR::Register* result  = cg->allocateRegister();
+   auto address = cg->evaluate(node->getChild(0));
+   auto value   = cg->gprClobberEvaluate(node->getChild(1), MOVRegReg());
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Sure, updated.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 4]
```diff
@@ -3009,27 +3065,43 @@ TR::Register *OMR::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::C
...
+         {
+         return TR::Compiler->target.is64Bit() ? inlineAtomicCompareAndMemoryUpdate(node, LCMPXCHG8MemReg, false, cg) : inline64BitAtomicCompareAndMemoryUpdateOn32Bit(node, false, cg);
+         }
+      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicCompareAndSwapReturnValue32BitSymbol))
+         {
+         return inlineAtomicCompareAndMemoryUpdate(node, LCMPXCHG4MemReg, true, cg);
+         }
+      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicCompareAndSwapReturnValue64BitSymbol))
+         {
+         return TR::Compiler->target.is64Bit() ? inlineAtomicCompareAndMemoryUpdate(node, LCMPXCHG8MemReg, true, cg) : inline64BitAtomicCompareAndMemoryUpdateOn32Bit(node, false, cg);
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Good catch, fixed.

File: compiler/x/codegen/OMRTreeEvaluator.cpp (hunk 1 above)
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Sure, updated to `TR::Register*`

**Comment by 0dvictor** [comment]
> My point about testing is that I would like to have some regular means of testing this in OMR CI builds (e.g., test cases).

> @0dvictor ping. I'd appreciate it if you could spend some time helping us to get through these older PRs. Please address @0xdaryl 's review comments.

Sorry that I missed the conversation and thank you for reminding. I'll look into prepare test cases and update the PR as soon as possible.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 5]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
   return node->getRegister();
 }
 
-TR::Register* OMR::X86::TreeEvaluator::performSimpleAtomicMemoryUpdate(TR::Node* node, TR_X86OpCodes op, TR::CodeGenerator* cg)
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Updated.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 6]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
...
+   generateMemRegInstruction(op, node, generateX86MemoryReference(address, 0, cg), value, cg);
 
-   node->setRegister(result);
+   node->setRegister(value);
    cg->decReferenceCount(node->getChild(0));
    cg->decReferenceCount(node->getChild(1));
-   return result;
+   return value;
+   }
+
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Updated.

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 7]
```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
...
+      }
+
+   node->setRegister(oldvalue);
+   cg->decReferenceCount(node->getChild(0));
+   cg->decReferenceCount(node->getChild(1));
+   cg->decReferenceCount(node->getChild(2));
+   return oldvalue;
+   }
+
+static TR::Register* inline64BitAtomicCompareAndMemoryUpdateOn32Bit(TR::Node* node, bool returnVaue, TR::CodeGenerator* cg)
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: Updated.

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@0xdaryl FYI this is still marked as WIP. @0dvictor are we waiting on anything?

**Comment by 0dvictor** [comment]
> are we waiting on anything?

Although these changes have been verified via OpenJ9's tests, there is not any OMR test cases. Unfortunately OMR's existing test framework does not allow tests on calls with "non-helper" intrinsic. As committer requested test cases, therefore it is marked as WIP. I'll remove the WIP when such test framework is ready.

**Comment by liqunl** [comment]
@0dvictor @0xdaryl The change in this PR is important to openj9 VarHandle performance, when can this PR be merged?

**Comment by Leonardo2718** [comment]
Change is still marked as WIP, so presumably it's not ready to be merged yet.

**Comment by 0dvictor** [comment]
The changeset itself is ready and has been tested via OpenJ9 prototypes; however, on OMR side, there is no unit tests as current OMR test frameworks do not allow set a "Nonhelper" symbol on `TR::xcall` opcode. @Leonardo2718 do you know how hard to add support for testing "Nonhelper" on OMR?

**Comment by Leonardo2718** [comment]
I think adding support for setting nonhelper symbols in Tril *should be* straight forward. Essentially, the code that handles call opcodes has to be modified to detect a pattern like `(call nonhelper="<atomicCompareAndSwapReturnValue>" ...)` and then use the information from `TR_Debug` to figure out what symbol to use. 

What I'm unsure about is how to verify the result of the test. It would be easy to just run the generated code but that won't check if it's atomic.

**Comment by 0xdaryl** [comment]
@0dvictor:  Please open an issue to track the creation of test cases for this feature and to capture some of the testing suggestions above.  The code changes themselves look fine, and the lack of OMR-specific tests is inline with other similar helpers for which tests do not exist.  Although not an ideal situation, I don't think merging needs to be held up because it.

**Comment by 0dvictor** [comment]
Good point, I have created https://github.com/eclipse/omr/issues/3758 to track the enhancement of test framework so that related tests of this pull requests can be introduced.

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Pull request merged**